mypy/fastparse.py

```diff
@@ -213,10 +214,64 @@ def parse(
         assert options.python_version[0] >= 3
         feature_version = options.python_version[1]
     try:
+        ignore_comment_errors = options and options.ignore_comment_errors
+
+        def try_parse_ast(source: str | bytes) -> tuple[AST | None, str | bytes]:
```

Comment:
I'd move this out as a top-level private function (start with underscore) - something like `_ast3_parse_ignore_bad_type_comments`:
- it will reduce indentation
- it will make the `parse` function read clearer, i.e.
```python
if options.ignore_comment_errors:
  ast, source = _ast3_parse_ignore_bad_type_comments(...)
else:
  ast, source = ast3_parse(...)
```

The `while` loop will be within the `_ast3_parse_ignore_bad_type_comments` function itself.

Reply:
Agreed. I don't think it matters for this module, but do you recommend putting underscore functions at the beginning of the module, or near the functions where they are first used?

Reply:
IMO, near the function. I personally like locality.

Reply:
I moved the re-parsing out to its own function, like you suggested. I made it recursive, which I think is cleaner. It also preserves the original exception and raises that if the source can't be parsed, even after tweaking.

The interesting thing about this reparse function is that it could be expanded upon to handle any parsing error that you wanted to try to recover from, based on the options passed.

Reply:
I tried to use `AnyStr` to avoid the code duplication. It passed locally, but is failing in CI. I'll look at that when I get a chance.